proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,136 @@
+* Proposal: [0010](0010-resource-element-type-validation.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: Joshua Batista
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`. The
+resource type is not a `RawBuffer` variant, and so there is a distinct set
+of rules that define valid RETs for this resource type.
+
+RETs for non-`RawBuffer` variants may include basic types (ints and uints of sizes 16
+and 32, as well as half, float, vectors, and matrices of 4 elements or fewer).
+Structs that contain fields of these basic types (where all fields in the struct have
+the same type) may also be RETs. 
+Structs that either have structs as fields or arrays of structs as fields may also be
+allowed, as long as there are at most 4 sub elements, and each sub element is at most
+32 bits. Additionally, resource types are not allowed within an RET, even if the 
+underlying resource type has a primitive RET (i.e., `RWBuffer<int>` as an RET).
+
+RETs for `RawBuffer` variants are much less constrained, the only rule is that the RET
+may not be an incomplete type (a handle type or a resource type).
+
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not an int, uint, nor a floating-point type, or vectors 
+or matrices containing the aforementioned types, will be rejected. The allow list isn't
+broad enough, because it doesn't include the case where the RET is user-defined. 
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 

```

Comment:
> a user should be able to determine how any user-defined structure is invalid as an RET

Is this a statement about the type of diagnostics we need to emit, or something else?

Reply:
Yes, this is a statement highlighting the lack of detailed diagnostics emitted when an invalid UDT is used as an RET.
We want to emit detailed diagnostics, but primarily, emit *some* diagnostic, in the UDT as RET case. It gives motivation for this spec.